### ISSUE TITLE
I think it's possible to enforce that a password is required on the User model without leaking it.

### ISSUE THREAD

**Issue by alejandrorpascual** [epicweb-dev/web-auth#7]
In `exercise 4` you wrote:

> Unfortunately, it's not possible to enforce that a password is required on the User model at the database level. However, it's worth the tradeoff to avoid the risk of leaking passwords.

I was checking the `Prisma` docs for `one-to-one` relationships, and found that if you change the relationship scalar (or foreign key) you can make the password field required when creating a `User`

from `Prisma` docs ([Required and optional 1-1 relation fields](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/one-to-one-relations#required-and-optional-1-1-relation-fields))

```prisma
model User {
  id      Int      @id @default(autoincrement())
  profile Profile? // No relation scalar - must be optional
}
```

> However, you can choose if the side of the relation with a relation scalar should be optional or mandatory. In the following example, `profile` and `profileId` are mandatory. This means that you cannot create a `User` without connecting or creating a `Profile`:

```prisma
model User {
  id        Int     @id @default(autoincrement())
  profile   Profile @relation(fields: [profileId], references: [id]) // references `id` of `Profile`
  profileId Int     @unique // relation scalar field (used in the `@relation` attribute above)
}

model Profile {
  id   Int   @id @default(autoincrement())
  user User?
}
```

so, if you do something like this:

```prisma
model Password {
  id   String @id @default(cuid())
  hash String
  // 👇 Now this is optional
  user User?
}


model User {
  id       String  @id @default(cuid())
  email    String  @unique
  username String  @unique
  name     String?

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  image      UserImage?
  // 👇 note the fields here.
  password   Password   @relation(fields: [passwordId], references: [id])
  passwordId String     @unique

  notes Note[]
}
```
The `password` would be required for `User` creation, and I think it won't eagerly load the relationship field (`password` on the `User` model in this case) because `Prisma` doesn't load those by default.

**Comment by kentcdodds** [comment]
Good point! Unfortunately this just moves the problem: now you can't enforce the user is required on the password. This was my intent with what I said. You can't enforce this to be required on both sides of the relationship.

But what you've said could be added as a note in the exercise instructions. Would you be interested in making a contribution? It could even be just a link to this issue.

**Comment by kentcdodds** [comment]
Thanks for [the help](https://github.com/epicweb-dev/web-auth/commit/37f9834949c53ed743365b5959483bffed283e26)!

**Comment by updbqn** [comment]
A suggestion for a potential future improvement since this issue is linked from the exercise.

Excluding a field globally by default, i.e. don't include it if not explicitly requested, is available as a preview feature since Prisma 5.16.0 (preview => mostly stable but not recommended for production yet).

- [Excluding fields | Prisma Documentation](https://www.prisma.io/docs/orm/prisma-client/queries/excluding-fields#excluding-a-field-globally-using-omit)

```prisma
// prisma/schema.prisma

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["omitApi"] // 👈 enable preview feature
}

model User {
  //...
  hash  String // 👈 Store hash directly on User
}
```

```typescript
// app/utils/db.server.ts

const prisma = new PrismaClient({
  omit: { user: { hash: true } }, // 👈 omit hash globally
});
```

```typescript
// hash is omitted by default
const user = await prisma.user.findUnique({ where: { id: 1 } });

// explicitly include hash
const user = await prisma.user.findUnique({
  select: { username: true, hash: true },
  where: { id: 1 },
});
// or
const user = await prisma.user.findUnique({
  omit: { hash: false },
  where: { id: 1 },
});

```

**Comment by kentcdodds** [comment]
Thanks for that note!

For myself I think I still prefer to keep them separate anyway to keep things clean even if accessing data without Prisma.
